### ISSUE TITLE
compiler warning about use of snprintf

### ISSUE THREAD

**Issue by tru7** [libesmtp/libESMTP#5]
One of my compilers (gcc (Debian 8.3.0-6) 8.3.0) complains about use of snprintf

smtp-tls.c: In function ‘starttls_create_ssl’:
smtp-tls.c:124:44: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2033 [-Wformat-truncation=]
   snprintf (buf, buflen, "%s/.authenticate/%s", home, tail);
                                            ^~
smtp-tls.c:369:45:
   keyfile = user_pathname (buf, sizeof buf, buf2);
                                             ~~~~
smtp-tls.c:124:3: note: ‘snprintf’ output 16 or more bytes (assuming 2063) into a destination of size 2048
   snprintf (buf, buflen, "%s/.authenticate/%s", home, tail);

I found that this could be avoided by dealing with the return value of snprintf. I'm not a developper so I really don't know how to correctly react if the length of pathname indeed exceeds 2048 characters in this situation. However this warning could be avoided by simply using this (example found by research):
snprintf (buf, buflen, "%s/.authenticate/%s", home, tail) < 0 ? abort() : (void)0;

Well, it's just a warning, but there is another aspect: if maximum length for pathname in linux is set to 4096 wouldn't it make sense to set buf size just to max, namely:
  char buf[PATH_MAX];
  char buf2[PATH_MAX];
in "static SSL_CTX" and "static SSL" 

would you accept a pull request for this?

**Comment by libesmtp** [comment]
I'm surprised I didn't spot that myself!  I've seen this warning quite a few times recently in other projects with recent GCC. I'll investigate further as I'd rather just fail the API in the event of truncation rather than abort(), or better still allocate a sufficiently long buffer.

**Comment by tru7** [comment]
Well, abort() does not happen when truncation happens. Here abort() only happens when there is an error which is not handled by now anyway (and actually cannot happen anyway). So the code change wouldn't do anything worse. When truncation happens snprintf just returns the number of characters actually taken. Buffer length should be sufficient with PATH_MAX.

My intention is to apply this fix to branch libesmtp-1.0.6 in order to achieve a decent status of old code for a new release soon.

**Comment by libesmtp** [comment]
Certain versions of snprintf() can fail and set errno according to https://pubs.opengroup.org/onlinepubs/9699919799/functions/fprintf.html however printf() functions are permitted fail for other reasons.

If a pathname is truncated, usually the worst than can happen is that the application cannot present a certificate and authentication fails as a result. However, truncation may cause a different file to be referenced making this a potential exploit (e.g. stuff the pathname with repeated `./` to force overflow) with a carefully crafted pathname.

If failure occurs, however unlikely, a library should never, on principle, call abort() as this would pull the rug from under the application - which may have a way to handle failure gracefully.  So I can't accept a fix that calls abort().

It isn't specified what the buffer passed to snprintf() might contain if it returns < 0 so the current code may segfault or similar anyway and that is unacceptable.  So I'm going to consider this issue a real bug and deal with it accordingly.

In this case the solution is to check failure or truncation and return a path to /dev/null, but the code needs review for other instances of this kind of thing.

**Comment by tru7** [comment]
Ok, this sounds very reasonable to me. I appreciate that you take it seriously. I would have no clue how to deal with and I don't have any idea what kind of outcome a call to abort() can have.  Honestly I just copied an example code. That's why I opened an issue to discuss it instead of just starting with a pull request

**Comment by libesmtp** [comment]
abort() deliberately kills the program by throwing SIGABRT which is a fatal signal.

**Comment by libesmtp** [comment]
Resolved with v1.1.0 release
